### PR TITLE
Support WeNet to run on Graphcore IPU

### PR DESCRIPTION
**Pull request by biofoolgreen**
First commit to support Graphcore IPU.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -14,3 +14,4 @@ flake8-pyi==20.5.0
 mccabe
 pycodestyle==2.6.0
 pyflakes==2.2.0
+https://download.pytorch.org/whl/cpu/torchaudio-0.10.0%2Bcpu-cp36-cp36m-linux_x86_64.whl
```
Comment (robin1001) [inline requirements.txt:17]: torchaudio is installed by conda, so it is not required here.

File: wenet/bin/recognize.py [hunk 2]
```diff
@@ -128,10 +128,10 @@ def main():
     symbol_table = read_symbol_table(args.dict)
     test_conf = copy.deepcopy(configs['dataset_conf'])
 
-    test_conf['filter_conf']['max_length'] = 102400
-    test_conf['filter_conf']['min_length'] = 0
-    test_conf['filter_conf']['token_max_length'] = 102400
-    test_conf['filter_conf']['token_min_length'] = 0
+    test_conf['filter_conf']['max_length'] = 1200
```
Comment (robin1001) [inline wenet/bin/recognize.py:131]: why change the default parameter here? is there any limits on IPU?

File: wenet/bin/train.py [hunk 3]
```diff
@@ -243,7 +282,11 @@ def main():
         device = torch.device('cuda' if use_cuda else 'cpu')
         model = model.to(device)
 
-    optimizer = optim.Adam(model.parameters(), **configs['optim_conf'])
+    optimizer = poptorch.optim.Adam(
+        model.parameters(),
+        accum_type=torch.float16,
```
Comment (robin1001) [inline wenet/bin/train.py:287]: condational set it  to torch.float16?

File: wenet/dataset/dataset.py [hunk 4]
```diff
@@ -187,3 +187,31 @@ def Dataset(data_type,
...
+        self.max_target_length = max_target_length
+        assert type in ['float32', 'float16']
+        self.type = type
+
+    def __call__(self, batch):
+        batch_size = len(batch)
+        feature_length = torch.cat([i[3] for i in batch])
+        target_length = torch.cat([i[4] for i in batch])
+        padded_feature = torch.zeros(
+            size=[batch_size, self.max_feature_length, batch[0][1].size(2)])
```
Comment (robin1001) [inline wenet/dataset/dataset.py:208]: What's the difference of this code compared to the pading function in processor.py?

File: wenet/transformer/asr_model.py [hunk 5]
```diff
@@ -718,6 +727,20 @@ def init_asr_model(configs):
         decoder = BiTransformerDecoder(vocab_size, encoder.output_size(),
                                        **configs['decoder_conf'])
     ctc = CTC(vocab_size, encoder.output_size())
+
```
Comment (robin1001) [inline wenet/transformer/asr_model.py:730]: If it is caused by torch.jit.export? we can simply remove jit check in train.py.

File: wenet/transformer/convolution.py [hunk 6]
```diff
@@ -66,7 +66,7 @@ def __init__(self,
             self.norm = nn.BatchNorm1d(channels)
         else:
             self.use_layer_norm = True
-            self.norm = nn.LayerNorm(channels)
+            self.norm = nn.LayerNorm(channels, eps=1e-3)
```
Comment (robin1001) [inline wenet/transformer/convolution.py:69]: any explanation or comment about setting it to 1e-3?

File: wenet/transformer/label_smoothing_loss.py [hunk 7]
```diff
@@ -48,11 +48,12 @@ def __init__(self,
                  normalize_length: bool = False):
         """Construct an LabelSmoothingLoss object."""
         super(LabelSmoothingLoss, self).__init__()
-        self.criterion = nn.KLDivLoss(reduction="none")
+        self.size = size
```
Comment (robin1001) [inline wenet/transformer/label_smoothing_loss.py:51]: KLD is not support on IPU?

Comment (tongjinle123) [reply wenet/transformer/label_smoothing_loss.py:51]: we have rewritten a label smoothing loss, and they are tested to be equal,
the original code's scatter op have some problem in performance.

File: wenet/transformer/convolution.py (hunk 6 above)
Comment (tongjinle123) [reply wenet/transformer/convolution.py:69]: our model is mainly trained on fp16, and the precision is set to 1e-3 to prevent numerical problem

File: wenet/transformer/asr_model.py (hunk 5 above)
Comment (tongjinle123) [reply wenet/transformer/asr_model.py:730]: yes, I make such walk around rather than just remove the jit because I found those exported function is used in runtime/, 
could be removed if it is ok.

File: wenet/dataset/dataset.py (hunk 4 above)
Comment (tongjinle123) [reply wenet/dataset/dataset.py:208]: the padding shape is statics while the original function is dynamic

File: wenet/bin/recognize.py (hunk 2 above)
Comment (tongjinle123) [reply wenet/bin/recognize.py:131]: is reset to default, a wrong change in the very beginning

File: requirements.txt (hunk 1 above)
Comment (tongjinle123) [reply requirements.txt:17]: our SDK environment is using python3.6, and the installation steps may have to follow the IPU_README.md.

there is also a problem for the checking below: `UnitTest / unit-test (ubuntu-latest, 3.8, 1.10.0, 0.10.0) (pull_request) Failing after 2m — unit-test (ubuntu-latest, 3.8, 1.10.0, 0.10.0)`

File: wenet/bin/train.py (hunk 3 above)
Comment (tongjinle123) [reply wenet/bin/train.py]: modified to follow the `precision`
